libnavigation-model/src/main/java/com/mapbox/navigation/model/MetricsRouteProgress.kt

```diff
@@ -0,0 +1,136 @@
+    var legCount: Int = 0
+        private set
+    var stepIndex: Int = 0
+        private set
+    var stepCount: Int = 0
+        private set
+
+    init {
+        ifNonNull(
+            routeProgress, routeProgress?.directionsRoute(),
+            routeProgress?.currentLegProgress(),
+            routeProgress?.distanceRemaining(), routeProgress?.durationRemaining()
+        ) { _routeProgress, _directionsRoute, _currentLegProgress, _distanceRemaining, _durationRemaining ->
+            obtainRouteData(_directionsRoute)
+            obtainLegData(_currentLegProgress)
+            obtainStepData(_routeProgress)
+            distanceRemaining = _distanceRemaining.toInt()
+            durationRemaining = _durationRemaining.toInt()
+            distanceTraveled = _routeProgress.distanceTraveled()?.toInt() ?: 0
+            legIndex = _routeProgress.legIndex() ?: 0
+            legCount = _directionsRoute.legs()?.size ?: 0
+            stepIndex = _currentLegProgress.stepIndex() ?: 0
+            stepCount = _routeProgress.currentLeg()?.steps()?.size ?: 0
+        } ?: initDefaultValues()
+    }
+
+    private fun initDefaultValues() {
+        directionsRouteProfile = ""
+        directionsRouteDestination = DEFAULT_POINT
+        currentStepName = ""
+        upcomingStepInstruction = ""
+        upcomingStepModifier = ""
+        upcomingStepType = ""
+        upcomingStepName = ""
+        previousStepInstruction = ""
+        previousStepModifier = ""
+        previousStepType = ""
+        previousStepName = ""
+    }
+
+    private fun obtainRouteData(route: DirectionsRoute) {
+        directionsRouteDistance = route.distance()?.toInt() ?: 0
+        directionsRouteDuration = route.duration()?.toInt() ?: 0
+        directionsRouteProfile = route.routeOptions()?.profile() ?: ""
+        directionsRouteDestination = retrieveRouteDestination(route)
+    }
+
+    private fun obtainLegData(legProgress: RouteLegProgress) {
+        currentStepDistance = legProgress.currentStep()?.distance()?.toInt() ?: 0
+        currentStepDuration = legProgress.currentStep()?.duration()?.toInt() ?: 0
+        currentStepDistanceRemaining =
+            legProgress.currentStepProgress()?.distanceRemaining()?.toInt() ?: 0
+        currentStepDurationRemaining =
+            legProgress.currentStepProgress()?.durationRemaining()?.toInt() ?: 0
+        currentStepName = legProgress.currentStep()?.name() ?: ""
+    }
+
+    private fun obtainStepData(routeProgress: RouteProgress) {
+        val legProgress = routeProgress.currentLegProgress()
+        legProgress?.upComingStep()?.let { upcomingStep ->
+            upcomingStepName = upcomingStep.name()
+            upcomingStep.maneuver().let {
+                upcomingStepInstruction = it.instruction()
+                upcomingStepType = it.type()
+                upcomingStepModifier = it.modifier()
+            }
+        }
+        legProgress?.currentStep()?.maneuver().let { stepManeuver ->
```

Comment:
```suggestion
        legProgress?.currentStep()?.maneuver().?let { stepManeuver ->
```
or `stepManeuver` might be `null`?